src/t8_schemes/t8_scheme_common_cxx.hxx

```diff
@@ -0,0 +1,631 @@
+      t8_element_successor (elem2, elem2, level - 1);
+      t8_element_child (elem2, 0, elem2);
+    }
+    else {
+      t8_element_sibling (elem1, child_id + 1, elem2);
+    }
+  }
+  /*****/
+  virtual int
+  t8_element_ancestor_id (const t8_element_t *elem, int level) const override
+  {
+    T8_ASSERT (0 < level);
+    T8_ASSERT (level <= t8_element_level (elem));
+    t8_element_t *anc;
+    t8_element_new (1, &anc);
+    t8_element_copy (elem, anc);
+    while (t8_element_level (anc) > level) {
+      t8_element_parent (anc, anc);
+    }
+    int child_id = t8_element_child_id (anc);
+    t8_element_destroy (1, &anc);
+    return child_id;
+  }
+
+  virtual void
+  t8_element_nca (const t8_element_t *elem1, const t8_element_t *elem2, t8_element_t *nca) const override
+  {
+    t8_element_t *anc1;
+    t8_element_t *anc2;
+    t8_element_new (1, &anc1);
+    t8_element_new (1, &anc2);
+    t8_element_copy (elem1, anc1);
+    t8_element_copy (elem2, anc2);
+
+    while (t8_element_level (anc1) > t8_element_level (anc2))
+      t8_element_parent (anc1, anc1);
+    while (t8_element_level (anc1) < t8_element_level (anc2))
+      t8_element_parent (anc2, anc2);
+
+    while (!t8_element_equal (anc1, anc2)) {
+      t8_element_parent (anc1, anc1);
+      t8_element_parent (anc2, anc2);
+    }
+    t8_element_copy (anc1, nca);
+    t8_element_destroy (1, &anc1);
+    t8_element_destroy (1, &anc2);
+  }
+
+  virtual void
+  t8_element_children (const t8_element_t *elem, int length, t8_element_t *children[]) const override
+  {
+    T8_ASSERT (length == t8_element_num_children (elem));
+    for (int ichild = 0; ichild < length; ichild++) {
+      t8_element_child (elem, ichild, children[ichild]);
+    }
+  }
+
+  virtual int
+  t8_element_num_siblings (const t8_element_t *elem) const override
+  {
+    T8_ASSERT (t8_element_is_valid (elem));
+    t8_element_t *parent;
+    t8_element_new (1, &parent);
+    t8_element_parent (elem, parent);
+    return t8_element_num_children (parent);
+  }
+
+  virtual void
+  t8_element_sibling (const t8_element_t *elem, int sibid, t8_element_t *sibling) const override
+  {
+    T8_ASSERT (t8_element_is_valid (elem));
+    T8_ASSERT (t8_element_is_valid (sibling));
+    t8_element_parent (elem, sibling);
+    t8_element_child (sibling, sibid, sibling);
+  }
+
+  /** compute linear ids of both elems. Compare ids, if they are equal, compare levels*/
+  virtual int
+  t8_element_compare (const t8_element_t *elem1, const t8_element_t *elem2) const override
+  {
+    T8_ASSERT (t8_element_is_valid (elem1));
+    T8_ASSERT (t8_element_is_valid (elem2));
+    int level1 = t8_element_level (elem1);
+    int level2 = t8_element_level (elem2);
+    int maxlevel = SC_MAX (level1, level2);
+    t8_linearidx_t id1 = t8_element_get_linear_id (elem1, maxlevel);
+    t8_linearidx_t id2 = t8_element_get_linear_id (elem2, maxlevel);
+    return id1 < id2 ? -1 : (id1 > id2 ? 1 : (level1 < level2 ? -1 : (level1 > level2 ? 1 : 0)));
```

Comment:
Please document this expression. It is not that nice to look at and a comment might help